src/azul/__init__.py

```diff
@@ -825,12 +825,14 @@ def __attrs_post_init__(self):
             if self.internal:
                 assert self.is_integration_test_catalog is True, self
 
-            atlases = {
-                Plugin.atlas_for_type(type_name, plugin.name)
-                for type_name, plugin in self.plugins.items()
-            }
-            require(len(atlases) == 1,
-                    'Conflicting plugin atlases', atlases)
+            repository_plugin_name = self.plugins['repository'].name
```

Comment:
We try hard to avoid hard-coding the plugin type names. Imagine having to rename a plugin type: searches for `'meatdata'` and `'repository'` would likely yield many false positives, so that refactoring would tedious and error-prone.

```diff
Subject: [PATCH] [r] Fix: Partition sizing ignores supplementary bundles (#5207)
---
Index: src/azul/plugins/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/__init__.py b/src/azul/plugins/__init__.py
--- a/src/azul/plugins/__init__.py	(revision c6d873bfe5ba641359a0598dca1e5c0dea66a2cc)
+++ b/src/azul/plugins/__init__.py	(date 1684782900630)
@@ -155,9 +155,7 @@
 
         :param catalog: the name of the catalog for which to load the plugin
         """
-        assert cls != Plugin, f'Must use a subclass of {cls.__name__}'
-        assert isabstract(cls) != Plugin, f'Must use an abstract subclass of {cls.__name__}'
-        plugin_type_name = cls.type_name()
+        plugin_type_name = cls._plugin_type_name()
         plugin_package_name = config.catalogs[catalog].plugins[plugin_type_name].name
         return cls._load(plugin_type_name, plugin_package_name)
 
@@ -190,12 +188,20 @@
         raise NotImplementedError
 
     @classmethod
-    def bundle_cls(cls, plugin_type_name: str, plugin_package_name: str) -> Type[BUNDLE]:
+    def bundle_cls(cls, plugin_package_name: str) -> Type[BUNDLE]:
+        plugin_type_name = cls._plugin_type_name()
         plugin_cls = cls._load(plugin_type_name, plugin_package_name)
         bundle_cls, *_ = get_generic_type_params(plugin_cls)
         assert issubclass(bundle_cls, Bundle), bundle_cls
         return bundle_cls
 
+    @classmethod
+    def _plugin_type_name(cls):
+        assert cls != Plugin, f'Must use a subclass of {cls.__name__}'
+        assert isabstract(cls) != Plugin, f'Must use an abstract subclass of {cls.__name__}'
+        plugin_type_name = cls.type_name()
+        return plugin_type_name
+
     @classmethod
     def _load(cls, plugin_type_name: str, plugin_package_name: str) -> Type[T]:
         plugin_package_path = f'{__name__}.{plugin_type_name}.{plugin_package_name}'
Index: src/azul/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/__init__.py b/src/azul/__init__.py
--- a/src/azul/__init__.py	(revision c6d873bfe5ba641359a0598dca1e5c0dea66a2cc)
+++ b/src/azul/__init__.py	(date 1684783165652)
@@ -45,6 +45,10 @@
 from azul.json_freeze import (
     freeze,
 )
+from azul.plugins import (
+    MetadataPlugin,
+    RepositoryPlugin,
+)
 from azul.types import (
     JSON,
     LambdaContext,
@@ -820,19 +824,19 @@
             all_types = set(p.type_name() for p in Plugin.types())
             configured_types = self.plugins.keys()
             require(all_types == configured_types,
-                    'Missing or extra plugin types',
-                    all_types.symmetric_difference(configured_types))
+                    # REVIEW: Separate commit for this hunk
+                    'Catalog is missing or has extra plugin types',
+                    self.name, all_types.symmetric_difference(configured_types))
             if self.internal:
                 assert self.is_integration_test_catalog is True, self
 
-            repository_plugin_name = self.plugins['repository'].name
-            repository_bundle_cls = Plugin.bundle_cls('repository', repository_plugin_name)
-
-            metadata_plugin_name = self.plugins['metadata'].name
-            metadata_bundle_cls = Plugin.bundle_cls('metadata', metadata_plugin_name)
+            repository_bundle_cls, metadata_bundle_cls = (
+                plugin_type.bundle_cls(self.plugins[plugin_type.type_name()].name)
+                for plugin_type in [RepositoryPlugin, MetadataPlugin]
+            )
             require(issubclass(repository_bundle_cls, metadata_bundle_cls),
-                    'Configured plugins have incompatible metadata schemas',
-                    repository_plugin_name, metadata_plugin_name)
+                    'Catalog combines incompatible metadata and repository plugins',
+                    self.name, repository_bundle_cls, metadata_bundle_cls)
 
         @cached_property
         def is_integration_test_catalog(self) -> bool:
```